lib/src/simple_tokenizer.dart

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:angular_ast/src/simple_token.dart';
+import 'package:charcode/charcode.dart';
+import 'package:string_scanner/string_scanner.dart';
+import 'package:meta/meta.dart';
+
+class NgSimpleTokenizer {
+  @literal
+  const factory NgSimpleTokenizer() = NgSimpleTokenizer._;
+
+  const NgSimpleTokenizer._();
+
+  Iterable<NgSimpleToken> tokenize(String template) sync* {
+    final scanner = new NgSimpleScanner(template);
+    NgSimpleToken token = scanner.scan();
+    while (token.type != NgSimpleTokenType.EOF) {
+      yield token;
+      token = scanner.scan();
+    }
+  }
+}
+
+class NgSimpleScanner {
+  String get state =>
+      _state == _NgSimpleScannerState.element ? "element" : "text";
+  String get remainingString => _scanner.rest;
+  String get elementRegex => _allElementMatches.toString();
+  static bool matchesGroup(Match match, int group) =>
+      match.group(group) != null;
+
+  static final _allTextMatches = new RegExp(r'([^\<]+)|(<!--)|(<)');
+  static final _allElementMatches = new RegExp(r'(\])|' //1  ]
+      r'(\!)|' //2  !
+      r'(\-)|' //3  -
+      r'(\))|' //4  )
+      r'(>)|' //5  >
+      r'(\/)|' //6  /
+      r'(\[)|' //7  [
+      r'(\()|' //8  (
+      r'([\s]+)|' //9 whitespace
+      r'([a-zA-Z0-9]([a-zA-Z0-9\-\_]*[a-zA-Z0-9])?)|' //10 any alphanumeric + '-' + '_'
+      r'("([^"\\]|\\.)*")|' //12 closed double quote (includes group 13)
+      r"('([^'\\]|\\.)*')|" //14 closed single quote (includes group 15)
+      r'(")|' //16 " (floating)
+      r"(')|" //17 ' (floating)
+      r"(<)|" //18 <
+      r"(=)|" //19 =
+      r"(\*)"); //20 *
+  static final _commentEnd = new RegExp('-->');
+
+  final StringScanner _scanner;
+  _NgSimpleScannerState _state = _NgSimpleScannerState.text;
+
+  factory NgSimpleScanner(String html, {sourceUrl, initialTextState: true}) {
+    return new NgSimpleScanner._(new StringScanner(html, sourceUrl: sourceUrl),
+        initialTextState: initialTextState);
+  }
+
+  NgSimpleScanner._(this._scanner, {initialTextState})
+      : _state = (initialTextState)
+            ? _NgSimpleScannerState.text
+            : _NgSimpleScannerState.element;
+
+  NgSimpleToken scan() {
+    switch (_state) {
+      case _NgSimpleScannerState.comment:
+        return scanComment();
+      case _NgSimpleScannerState.commentEnd:
+        return scanCommentEnd();
+      case _NgSimpleScannerState.element:
+        return scanElement();
+      case _NgSimpleScannerState.text:
+        return scanText();
+    }
+    return null;
+  }
+
+  NgSimpleToken scanComment() {
+    int offset = _scanner.position;
+    while (true) {
+      if (_scanner.peekChar() == $dash &&
```

Comment:
I've initially done this as a regex, but there were too many edge cases where the regex would fail. A regex that worked would force me to return comment texts into multiple token instead of one single text token; this method seemed to yield the best accurate results.